arch/sim/src/sim/up_hcisocket.c

```diff
@@ -0,0 +1,235 @@
+ * Included Files
+ ****************************************************************************/
+
+#include <sys/types.h>
+
+#include <sys/socket.h>
+
+#include <stdbool.h>
+#include <stdio.h>
+#include <string.h>
+#include <unistd.h>
+#include <errno.h>
+
+#include <nuttx/wireless/bluetooth/bt_driver.h>
+#include <nuttx/net/bluetooth.h>
+
+#include "up_internal.h"
+#include "up_hcisocket_host.h"
+
+/****************************************************************************
+ * Pre-processor Definitions
+ ****************************************************************************/
+
+/* HCI data types as defined by Linux Kernel */
+
+#define HCI_COMMAND_PKT   0x01
+#define HCI_ACLDATA_PKT   0x02
+#define HCI_SCODATA_PKT   0x03
+#define HCI_EVENT_PKT     0x04
+#define HCI_ISODATA_PKT   0x05
+#define HCI_DIAG_PKT      0xf0
+#define HCI_VENDOR_PKT    0xff
+
+/****************************************************************************
+ * Private Types
+ ****************************************************************************/
+
+/****************************************************************************
+ * Private Function Prototypes
+ ****************************************************************************/
+
+static int bthcisock_send(FAR const struct bt_driver_s *dev,
+                          FAR struct bt_buf_s *buf);
+static int bthcisock_open(FAR const struct bt_driver_s *dev);
+
+/****************************************************************************
+ * Private Data
+ ****************************************************************************/
+
+static const struct bt_driver_s g_bt_hcisock =
+{
+  0,               /* head_reserve */
+  bthcisock_open,  /* open */
+  bthcisock_send   /* send */
+};
+
+static int bt_fd = -1;       /* Host HCI socket fd */
+static int host_dev_id = 0;  /* Host HCI interface number */
+
+/* This frame buffer is used to handle the frame read from the host
+ * NOTE: This is smaller than what the Linux kernel lists as its max frame
+ * so we will need to resolve that at some point.
+ */
+
+static uint8_t bt_frame[BLUETOOTH_MAX_FRAMELEN + 1];
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+static int bthcisock_send(FAR const struct bt_driver_s *dev,
+                          FAR struct bt_buf_s *buf)
+{
+  uint8_t pkt_type;
+
+  switch (buf->type)
+    {
+      case BT_CMD:
+        {
+          pkt_type = HCI_COMMAND_PKT;
+          break;
+        }
+
+      case BT_ACL_OUT:
+        {
+          pkt_type = HCI_ACLDATA_PKT;
+          break;
+        }
+
+      default:
+        {
+          wlerr("Unexpected HCI packet type %d", buf->type);
+          return buf->len;
+        }
+    }
+
+  if (bthcisock_host_send(bt_fd, pkt_type, buf->data, buf->len) < 0)
+    {
+      return -1;
+    }
+
+  bthcisock_read();
```

Comment:
should we centralize the read to the idle loop?

Reply:
I added this here because for most HCI commands it immediately expects to be acknowledged by the controller so this captures that packet fast if it's ready.  But removing it probably won't change much. I'm fine either way.

Reply:
Since the poll in the idle loop can't be avoided in the current simulator architecure, it's better to keep the code in the idle loop but remove the code here to simplify the logic.

Reply:
Ok I removed it. Idle loop is fast anyway in the sim target.

Reply:
Removed this call, but the function cannot be static since the idle loop needs to call it and I don't want to move all that logic into up_idle.c

Reply:
Yes, you are right.